paperless: Fix a few papercuts

mweinelt opened the pull request:
###### Description of changes

Resolves a few papercuts I experienced while setting up paperless.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
nixos/paperless: Allow mbind syscall in paperless-web.services

After uploading a document through the webinterface I started seeing
it killed through the SYSBUS signal. Inspecting the call trace led me to
liblapack's memory allocator, that uses the mbind syscall on Linux.

mweinelt committed:
nixos/paperless: Use system timezone by default, if set

mweinelt committed:
nixos/paperless: Add pgsql via unix socket example

Finding out how to connect paperless to a PostgreSQL database via unix
sockets and peer authentication took me a few minutes, so leaving a hint
in the extraConfig example seems like a good idea to me.

Also remove unnecessary use of literalExpression for attribute set, it
is only required for complex values like functions or values that depend
on other values or packages.

mweinelt commented:
There is still the following error visible in the NixOS tests.

```
vm-test-run-paperless> subtest: Documents are consumed
vm-test-run-paperless> machine: waiting for success: (($(curl -u admin:admin -fs localhost:28981/api/documents/ | jq .count) == 2))
vm-test-run-paperless> machine # [   29.821756] paperless-ngx[1071]: [2022-09-04 12:21:28,810] [INFO] [paperless.consumer] Consuming doc.png
vm-test-run-paperless> machine # [   30.666513] paperless-ngx[1071]: [2022-09-04 12:21:29,655] [WARNING] [paperless.parsing.tesseract] Error while getting DPI from image /var/lib/paperless'
vm-test-run-paperless> machine # [   33.225031] paperless-ngx[1071]: [2022-09-04 12:21:32,213] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] Warning: Invalid resolution 48 dpi. Using 70.
vm-test-run-paperless> machine # [   33.226829] paperless-ngx[1071]: [2022-09-04 12:21:32,213] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] Warning. Invalid resolution 48 dpi. Using 70.
vm-test-run-paperless> machine # [   33.229178] paperless-ngx[1071]: [2022-09-04 12:21:32,213] [ERROR] [ocrmypdf._exec.tesseract] [tesseract] Error during processing.
vm-test-run-paperless> machine # [   34.208424] paperless-ngx[1071]: [2022-09-04 12:21:33,197] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] Warning: Invalid resolution 48 dpi. Using 70.
vm-test-run-paperless> machine # [   34.633477] paperless-ngx[1153]: convert: FailedToExecuteCommand `'gs' -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=5000000.
vm-test-run-paperless> machine # [   34.638537] paperless-ngx[1153]: convert: no images defined `/tmp/paperless/paperless-aid36n32/convert.webp' @ error/convert.c/ConvertImageCommand/3342.
vm-test-run-paperless> machine # [   34.656765] paperless-ngx[1071]: [2022-09-04 12:21:33,645] [WARNING] [paperless.parsing] Thumbnail generation with ImageMagick failed, falling back to g!
vm-test-run-paperless> machine # [   35.650601] paperless-ngx[1071]: [2022-09-04 12:21:34,639] [INFO] [paperless.consumer] Document 2005-10-16 doc consumption finished
vm-test-run-paperless> machine # [   35.653660] paperless-ngx[1071]: 12:21:34 [Q] INFO Process-1:1 stopped doing work
vm-test-run-paperless> machine # [   35.671751] paperless-ngx[1073]: 12:21:34 [Q] INFO Processed [doc.png]
vm-test-run-paperless> machine # [   38.267439] paperless-ngx[1070]: 12:21:37 [Q] INFO recycled worker Process-1:1
vm-test-run-paperless> machine # [   38.270700] paperless-ngx[1180]: 12:21:37 [Q] INFO Process-1:4 ready for work at 1180
vm-test-run-paperless> machine # [   38.271560] paperless-ngx[1180]: 12:21:37 [Q] INFO Process-1:4 processing [webdoc.png]
vm-test-run-paperless> machine # [   38.890814] paperless-ngx[1180]: [2022-09-04 12:21:37,879] [INFO] [paperless.consumer] Consuming webdoc.png
vm-test-run-paperless> machine # [   39.224800] paperless-ngx[1180]: [2022-09-04 12:21:38,213] [WARNING] [paperless.parsing.tesseract] Error while getting DPI from image /tmp/paperless/pap'
vm-test-run-paperless> machine # [   40.001992] paperless-ngx[1180]: [2022-09-04 12:21:38,990] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] Warning: Invalid resolution 48 dpi. Using 70.
vm-test-run-paperless> machine # [   40.002783] paperless-ngx[1180]: [2022-09-04 12:21:38,990] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] Warning. Invalid resolution 48 dpi. Using 70.
vm-test-run-paperless> machine # [   40.004080] paperless-ngx[1180]: [2022-09-04 12:21:38,990] [ERROR] [ocrmypdf._exec.tesseract] [tesseract] Error during processing.
vm-test-run-paperless> machine # [   40.756717] paperless-ngx[1180]: [2022-09-04 12:21:39,745] [WARNING] [ocrmypdf._exec.tesseract] [tesseract] Warning: Invalid resolution 48 dpi. Using 70.
vm-test-run-paperless> machine # [   41.041213] paperless-ngx[1224]: convert: FailedToExecuteCommand `'gs' -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=5000000.
vm-test-run-paperless> machine # [   41.044443] paperless-ngx[1224]: convert: no images defined `/tmp/paperless/paperless-h13cqq46/convert.webp' @ error/convert.c/ConvertImageCommand/3342.
vm-test-run-paperless> machine # [   41.066114] paperless-ngx[1180]: [2022-09-04 12:21:40,054] [WARNING] [paperless.parsing] Thumbnail generation with ImageMagick failed, falling back to g!
vm-test-run-paperless> machine # [   41.985678] paperless-ngx[1180]: [2022-09-04 12:21:40,974] [INFO] [paperless.consumer] Document 2005-10-16 webdoc consumption finished
vm-test-run-paperless> machine # [   41.988824] paperless-ngx[1180]: 12:21:40 [Q] INFO Process-1:4 stopped doing work
vm-test-run-paperless> machine # [   41.998447] paperless-ngx[1073]: 12:21:40 [Q] INFO Processed [webdoc.png]
vm-test-run-paperless> (finished: waiting for success: (($(curl -u admin:admin -fs localhost:28981/api/documents/ | jq .count) == 2)), in 14.13 seconds)

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -181,6 +181,7 @@ in
       example = literalExpression ''
         {
           PAPERLESS_OCR_LANGUAGE = "deu+eng";
+          PAPERLESS_DBHOST = "/run/postgresql";
```
While you're at it, can you remove the unneeded `literalExpression`?

mweinelt replied:
I thought that was exactly where `literalExpression` was supposed to be used. What makes you think otherwise?

erikarvstedt replied:
Ah, sorry, I meant the extra string wrapping (`''`) of the attrset.

erikarvstedt replied:
`literalexpression` is for nix expressions that that can't be pretty-printed like `literalExpression "pkgs.hello"`.

mweinelt replied:
Checked the manual for good measure and it says that it is only required for complex values, where some kind of evaluation needs to happen.

erikarvstedt reviewed (changes requested):
Commit `paperless: Set paths to convert and gs executables` is unneeded.
If the variables are unset, paperless-ngx falls back to using `convert` and `gs`, and these are in PATH of the service.

mweinelt commented:
The documentation suggests that it is by default set to `/usr/bin/convert` and `/usr/bin/gs` respectively and thereby non-empty, so I guess it should be clarified if that is the case.

https://paperless-ngx.readthedocs.io/en/latest/configuration.html#binaries

erikarvstedt commented:
That's a documentation error, I'll fix it upstream.
Besides convert and gs, there are [many more binaries](https://github.com/NixOS/nixpkgs/blob/2c4c52179c4490cb88c9ca16074a75c03382c13c/pkgs/applications/office/paperless-ngx/default.nix#L52) that are picked up via `PATH`.

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -17,9 +17,11 @@ let
     lib.mapAttrs (_: toString) cfg.extraConfig
-  ) // (optionalAttrs enableRedis {
+  ) // optionalAttrs (config.time.timeZone != null) {
+    PAPERLESS_TIMEZONE = lib.mkDefault config.time.timeZone;
```
Must be `PAPERLESS_TIME_ZONE`.

mweinelt replied:
Good catch, fixed.

mweinelt committed:
paperless: Expose python environment in passthru

This allows adding more python dependencies through overrides.

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -288,7 +289,7 @@ in
         AmbientCapabilities = "CAP_NET_BIND_SERVICE";
         CapabilityBoundingSet = "CAP_NET_BIND_SERVICE";
         # gunicorn needs setuid
```
nit:
```suggestion
        # gunicorn needs setuid, liblapack needs mbind
```

mweinelt committed:
nixos/paperless: Restrict CAP_NET_BIND_SERVICE

Handing CAP_NET_BIND_SERVICE to the `paperless-web.service` only makes
sense when it actually wants to bind to a port < 1024. Don't hand it out
if that is not the case.

mweinelt commented:
@ofborg eval

NobbZ commented:
It seems as if commit 81a17f7352bd7acc438c298f317c190d2fbdcc35 from this PR somehow broke my setup.

```
error: cannot coerce a set to a string

       at /nix/store/wibagkzj3knj4269x289x0jx0kcmgf5c-source/nixos/modules/services/misc/paperless.nix:27:88:

           26|   manage = let
           27|     setupEnv = lib.concatStringsSep "\n" (mapAttrsToList (name: val: "export ${name}=\"${val}\"") env);
             |                                                                                        ^
           28|   in pkgs.writeShellScript "manage" ''
(use '--show-trace' to show detailed location information)
```

with `time.timeZone = "Europe/Berlin"`.
